drivers/haptic/power_relay.c

```diff
@@ -0,0 +1,125 @@
+/* Copyright 2018 mtdjr - modified by ishtob
+ * Driver for solenoid written for QMK
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "timer.h"
+#include "power_relay.h"
+#include "haptic.h"
+#include "gpio.h"
+#include "usb_device_state.h"
+#include "print.h"
+
+uint8_t      power_relay_dwell  = POWER_RELAY_DEFAULT_DWELL;
+static pin_t power_relay_pads[] = POWER_RELAY_PINS;
+#define NUMBER_OF_RELAYS (sizeof(power_relay_pads) / sizeof(pin_t))
+bool     power_relay_on[NUMBER_OF_RELAYS]      = {false};
+bool     power_relay_buzzing[NUMBER_OF_RELAYS] = {false};
+uint16_t power_relay_start[NUMBER_OF_RELAYS]   = {0};
+#define low false
+#define high true
+static bool power_relay_active_state[NUMBER_OF_RELAYS];
+
+extern haptic_config_t haptic_config;
+
+void power_relay_buzz_on(void) { haptic_set_buzz(1); }
+
+void power_relay_buzz_off(void) { haptic_set_buzz(0); }
+
+void power_relay_set_buzz(uint8_t buzz) { haptic_set_buzz(buzz); }
+
+void power_relay_set_dwell(uint8_t dwell) { power_relay_dwell = dwell; }
+
+void power_relay_stop(uint8_t index) {
+    writePin(power_relay_pads[index], !power_relay_active_state[index]);
+    power_relay_on[index]      = false;
+    power_relay_buzzing[index] = false;
+}
+
+void power_relay_fire(uint8_t index) {
+    if (!haptic_config.buzz && power_relay_on[index]) return;
+    if (haptic_config.buzz && power_relay_buzzing[index]) return;
+
+    power_relay_on[index]      = true;
+    power_relay_buzzing[index] = true;
+    power_relay_start[index]   = timer_read();
+    writePin(power_relay_pads[index], power_relay_active_state[index]);
+}
+
+void power_relay_fire_handler(void) {
+    for (uint8_t i = 0; i < NUMBER_OF_RELAYS; i++) {
+        if (!power_relay_on[i]) {
+            power_relay_fire(i);
+            break;
+        }
+    }
```

Comment:
I'd love to replace this with code to randomly select a relay, until an unused one is found, or exhausts all of the relays, but .... out of energy to figure out how to do so, ATM. 

For now, it just selects them sequentially.